### ISSUE TITLE
FP redundant_closure_call  with async closure

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#12174]
### Summary

.

### Lint Name

redundant_closure_call

### Reproducer

I tried this code:

```rust
#![feature(async_closure)]
fn foo() -> Box<dyn std::future::Future<Output = u32>> {
    let x = 0u32;
    Box::new((async || x)())
}

fn main() {
}
```

I saw this happen:

```
warning: function `foo` is never used
 --> src/main.rs:5:4
  |
5 | fn foo() -> Box<dyn std::future::Future<Output = u32>> {
  |    ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: try not to call a closure in the expression where it is declared
 --> src/main.rs:7:14
  |
7 |     Box::new((async || x)())
  |              ^^^^^^^^^^^^^^ help: try doing something like: `async { x }`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_call
  = note: `#[warn(clippy::redundant_closure_call)]` on by default

warning: `o` (bin "o" test) generated 2 warnings (run `cargo clippy --fix --bin "o" --tests` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `o`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0373]: async block may outlive the current function, but it borrows `x`, which is owned by the current function
 --> src/main.rs:7:14
  |
7 |     Box::new(async { x })
  |              ^^^^^^^^-^^
  |              |       |
  |              |       `x` is borrowed here
  |              may outlive borrowed value `x`
  |
note: async block is returned here
 --> src/main.rs:7:5
  |
7 |     Box::new(async { x })
  |     ^^^^^^^^^^^^^^^^^^^^^
help: to force the async block to take ownership of `x` (and any other referenced variables), use the `move` keyword
  |
7 |     Box::new(async move { x })
  |                    ++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0373`.
Original diagnostics will follow.
```

### Version

```text
rustc 1.77.0-nightly (25f8d01fd 2024-01-18)
binary: rustc
commit-hash: 25f8d01fd8bda339612d0c0a8844173a09205f7c
commit-date: 2024-01-18
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by y21** [comment]
Was it a compiler bug that the reproducer was even accepted in the first place? I'm not sure that it should compile, and it *is* a compile error now. As far as I know, capture modes should be the same for closures and async blocks, unless async closures work differently?

In any case, this isn't reproducible anymore.
